### ISSUE TITLE
trilinos, sundials, mfem breakage with superlu-dist@develop API change

### ISSUE THREAD

**Issue by balay** [xsdk-project/xsdk-issues#106]
cc: @liruipeng 

We've had this issue with petsc as-well - and the petsc code was updated for this

https://gitlab.com/petsc/petsc/-/commit/a7b55187c168a3c94455badbf0c9f954c0a5c4b9

This is likely to also  break superlu-dist usage from trilinos so cc: trilinos

[spack-build-out.txt](https://github.com/xsdk-project/xsdk-issues/files/4452918/spack-build-out.txt)

**Comment by balay** [comment]
Here is the log from trilinos build error with superlu-dist.

[spack-build-out.txt](https://github.com/xsdk-project/xsdk-issues/files/4452965/spack-build-out.txt)

Or perhaps I should file a different issue for trilinos+superlu-dist build

**Comment by ulrikeyang** [comment]
Hi Satish,
Did you try this with the most recent superlu_dist version. We just noticed that there has been a significant change in superlu_dist.
Thanks
Ulrike

**Comment by balay** [comment]
@ulrikeyang this is with superlu-dist@develop

This change is in superlu-dist@6.3.0 aswell

**Comment by balay** [comment]
Same issue with mfem+superlu-dist.

[spack-build-out.txt](https://github.com/xsdk-project/xsdk-issues/files/4453157/spack-build-out.txt)

**Comment by balay** [comment]
same issue with sundials

[spack-build-out.txt](https://github.com/xsdk-project/xsdk-issues/files/4453295/spack-build-out.txt)

**Comment by balay** [comment]
Bring back this issue. 

trilinos, sundials, mfem continue to be broken with current superlu-dist@master

**Comment by keitat** [comment]
> Here is the log from trilinos build error with superlu-dist.
> 
> [spack-build-out.txt](https://github.com/xsdk-project/xsdk-issues/files/4452965/spack-build-out.txt)
> 
> Or perhaps I should file a different issue for trilinos+superlu-dist build

What happens when amesos is disabled?  I did confirm Amesos2 produces the same type of error (correction!).

In trilinos/package.py
```
    variant('amesos',       default=True,
            description='Compile with Amesos')
    variant('amesos2',      default=True,
            description='Compile with Amesos2')
```

**Comment by balay** [comment]
If I disable amesos - I get:
```
dealii requires trilinos variant +amesos, but spec asked for ~amesos
```

cc: @bangerth

**Comment by keitat** [comment]
> If I disable amesos - I get:
> 
> ```
> dealii requires trilinos variant +amesos, but spec asked for ~amesos
> ```
> 
> cc: @bangerth

I did confirm that both amesos and amesos2 cannot handle the latest version.  I fixed my own copy.  Waiting for the approval of the pull request.

**Comment by keitat** [comment]
Trilinos provides SuperLU_DIST with Amesos and Amesos2.  We do not want to add a new capability to Amesos, a legacy package (Technically possible, but we encourage the users to switch to Amesos2).  If deal.II team is able to switch to Amesos2 (modern C++ direct solver interface) in a foreseeable future, that will be great.

**Comment by bangerth** [comment]
@keitat: I don't know which PR you refer to two messages above, but if I understand you correctly in your last comment, then you're saying that you're essentially abandoning Amesos instead of adding compatibility with newer SuperLU_dist? deal.II would then just be an innocent bystander death :-( 

That said, we don't reference Amesos in a lot of places. Basically, there are just two:
https://github.com/dealii/dealii/blob/master/source/lac/trilinos_solver.cc#L702-L743
where the `solver` variable has type `std::unique_ptr<Amesos_BaseSolver>`. This object is then used a few lines further down.

The other use is here, really just one line:
https://github.com/dealii/dealii/blob/master/source/lac/trilinos_precondition.cc#L631-L634

Do you happen to have a pointer how we could convert these two places to Amesos2?

**Comment by keitat** [comment]
@bangerth   I'd like to know how much deal.Ii uses Epetra, Amesos, etc (legacy packages of Trilinos).  If you use the legacy version of Trilinos heavily, this is going to be a big change that cannot be finished shortly.  Since these legacy packages does not have any plan to support future platforms and new libraries, I strongly recommend to have a roadmap to switch to the 2nd generation Trilinos packages (Tpetra, Amesos2,etc.).

**Comment by masterleinad** [comment]
We hardly support any Tpetra really (there is only a Tpetra vector wrapper). It's not reasonable to do this switch within a few weeks.

**Comment by bangerth** [comment]
@masterleinad is correct. We can switch from Amesos to Amesos2 if the latter supports Epetra, but we're set on Epetra for the moment. Switching to Tpetra is a long-term goal, but whenever we looked at that in the past, there were either packages that we need and that hadn't themselves switched to Tpetra, and/or Tpetra itself was still working through interface revisions that made it unattractive to switch to it.

**Comment by keitat** [comment]
> @masterleinad is correct. We can switch from Amesos to Amesos2 if the latter supports Epetra, but we're set on Epetra for the moment. Switching to Tpetra is a long-term goal, but whenever we looked at that in the past, there were either packages that we need and that hadn't themselves switched to Tpetra, and/or Tpetra itself was still working through interface revisions that made it unattractive to switch to it.

As a former user of PETSc, I had the same impression with the inconsistency of Tpetra in the past.  I am talking to the Trilinos leadership for possible short term solutions.

**Comment by bangerth** [comment]
@keitat Can you link to the patch you had that made Amesos(1) compatible with the latest SuperLU_dist? Maybe we can lend a lobbying hand?

**Comment by keitat** [comment]
@bangerth Let me prepare the new source in my Trilinos fork.  I will let you know.

**Comment by keitat** [comment]
This is one of my commit.  https://github.com/trilinos/Trilinos/commit/208643507e7deeb42e164defe0e04f5d5c686e23
Current PR does not include the changes in Amesos (I am proposing Amesos2 only).

**Comment by keitat** [comment]
@masterleinad @bangerth One possible solution is calling Amesos2 from Epetra (there is an adaptor APIs with some tester programs), but it only serves for a short term.   Belos, Tpetra, Amesos2 solution is getting more stable than the past, and they use Kokkos underneath for performance portability.  Despite the bitter experience, I am hoping you will consider them for long term solution (for interoperability and performance portability).

**Comment by bangerth** [comment]
Oh good grief, what a gratuitous change in SuperLU... Surely they could have left the old names as an alias to the new names. Or was this already deprecated years ago?

The changes you show in the patch are gratuitous, but neither large nor difficult. Why would Trilinos *not* want to add a corresponding patch to Amesos, instead rendering a widely used package unusable to lots of people? I get that they don't want to add new functionality, but a compatibility patch?

**Comment by xiaoyeli** [comment]
It's not gratuitous change intended to make life hard for everyone ...

We have got increasing requests from users to use both real and complex at
the same time, hence I have to rename some data structures and routine
names as precision-dependent (e.g., separating 'd' and 'z')

As we release the mixed-precision code in the future,  there will be 's'
and 'c' versions as well.  And perhaps 'sd', 'cz', etc.

Sherry

**Comment by bangerth** [comment]
@xiaoyeli -- yea, I should have phrased that in a less personal way. Sorry about that. It's been a long day already :-(

I think what I meant to say is this: I understand the need to differentiate based on data type, just as you do. But one could have left the old name (without the additional character) as an alias to the new name (with 'd' as additional character). This way, all of those codes that have no interest in using the broader set of underlying data types (or are just in maintenance mode, such as Amesos1) can continue to work as they always have, without the need to change anything.

**Comment by balay** [comment]
The current fix in petsc is adding this mapping. [for both d and z symbols]

https://gitlab.com/petsc/petsc/commit/a7b55187c168a3c94455badbf0c9f954c0a5c4b9

I guess that was good time (when this was added to petsc) to ask for this change in superlu-dist [instead of changing petsc].

[Likely the superlu-dist  release was already out - so we just fixed petsc]

If a fix is added now  to superlu-dist - We might have to revert this petsc change [or fix the version checks appropriately - this would require a new version of superlu-dist set  - even-though its not yet released ...]. I think this should be ok.

Note: the above petsc change is from 6 months ago - the superlu-dist change was earlier than that, and this issue was filed 5 months ago..

**Comment by bangerth** [comment]
The Trilinos patch mentioned earlier (https://github.com/trilinos/Trilinos/commit/208643507e7deeb42e164defe0e04f5d5c686e23) does essentially the same as the PETSc patch.

@balay -- strictly speaking, you don't actually have to undo your patch if something similar was put into SuperLU_dist: it's ok to re-`#define` a preprocessor symbol in PETSc code. So it would probably be *desirable* to remove the code again, but you wouldn't have to roll out a dot-release for it. It could wait for the next regular release.

**Comment by balay** [comment]
We (petsc) plan to have a new release in about a week from now [end of this month]

**Comment by xiaoyeli** [comment]
This type of alias does not work if the user wants to have both real and
complex *in one executable at the same time.   *That is, they want both
#include <superlu_ddefs.h>
#include <superlu_zdefs.h>
in the same code.  The old way seems to work for some compilers.  But at
least some compilers break, complaining conflicts or multiple
definitions, and cannot proceed (not just warnings).  That's why I split
into separate names.

As I mentioned, in the future, when we want to support more mixed
precision / mixed type codes, it will be even more messy.

**Comment by bangerth** [comment]
I don't think I understand. If you unconditionally have
```
  #define ScalePermstruct_t dScalePermstruct_t
  ...same for the other types...
```
then wouldn't that allow people to continue using the same names for types as they always have? Of course, if someone wants to use a mixture of types, they have to explicitly use `dScalePermstruct_t` and `zScalePermstruct_t`, but these codes have no backward compatibility issues anyway give that the `z*` types weren't available until recently anyway.

In other words, the symbol names without any prefix simply *always* refer to the `d*` types.

**Comment by balay** [comment]
I guess the issue is - its providing compatibility for half the users [or half the api] - i.e only 'd' version. i.e for superlu_ddefs and not for superlu_zdefs users.

For ex: PETSc uses superlu_ddefs for real build and superlu_zdefs for complex build - so would need the mapping for the other half - complex part  [as superlu_dist can provide this only for half the stuff -i.e real part]

And providing these macros for both superlu_ddefs superlu_zdefs won't work..

**Comment by balay** [comment]
BTW: superlu_zdefs is not new. It had the complex API using the same function names (as in superlu_ddefs). So one would switch from  superlu_ddefs to superlu_zdefs [same function names] for switching from real-superlu-dist to complex-superlu-dist.

**Comment by xiaoyeli** [comment]
Good explanation Satish :-)

In the past, ScalePermstruct_t was used in both superlu_ddefs.h, and
superlu_zdefs.h, i.e., supported both real and complex code. No conflict
occurs if you use real and complex *separately*.
In fact, some compilers even tolerate this kind of 'multiple definition'
when both real and complex are used at the same time. Recent years, we got
more and more bug reports about the conflict when real/complex are used
*together.*

In my view, it's not good to macro define ScalePermstruct_t to either 'd',
or 'z' at the superlu level.  Since superlu is at the bottom of the food
chain, it's better to leave this mapping decision to the upstream
libraries.  PETSc way is good, but I guess if petsc allows users to use
both real and complex, these macros cannot be used.

Sherry

**Comment by balay** [comment]
Perhaps its possible to provide the mapping this way:
- if only superlu_ddefs.h is included - then the foo -> dfoo mapping is available
- if only  superlu_zdefs.h is included - then the foo -> zfoo mapping is available
- if both are included - then somehow have only one mapping [the first include?]
  - or prevent both mapings to avoid corner cases? [don't know how this can be done].

But this makes the API more complex...

**Comment by bangerth** [comment]
@xiaoyeli: Oh, I finally understand! I had not considered that the same name is used differently depending on what header file you include. I'm not surprised that this yields errors -- the C and C++ standards both say that this is specifically not allowed :-) But I now understand the background of the change.

@balay Prevent both, this way:
```
// in ddefs.h:
#ifndef SUPERLU_SUPPRESS_ALIAS_DEFINITIONS
#  ifdef SUPERLU_ALIASES_DEFINED
#    error "You can't #include multiple superlu_*defs.h files into the same user file unless you #define SUPERLU_SUPPRESS_ALIAS_DEFINITIONS"
#  else
#     define SUPERLU_ALIASES_DEFINED
#     define ScalePermstruct_t dScalePermstruct_t
       ...same for the other types...
#  endif
#endif
```
and similarly in `zdefs.h`, just with the different alias to `zScalePermstruct_t`. This way, if you include only one of the files, you get the old behavior. If you include both `ddefs.h` and `zdefs.h`, you get an error, unless you define `SUPERLU_SUPPRESS_ALIAS_DEFINITIONS` before you include the two files, in which case the aliases are not defined at all.

Or you could turn it around and ask that uses `#define` a variable called `SUPERLU_DEFINE_ALIASES` in order to get the aliases. Though that's not backward compatible.

**Comment by balay** [comment]
>  If you include both ddefs.h and zdefs.h, you get an error, unless you define SUPERLU_SUPPRESS_ALIAS_DEFINITIONS

Sure  - but this is now placing the burden on the new preferred API [which is to allow using both] users to set this flag..

The alternative I could think off is - each include has:
```
#if defined(SUPERLU_ALIASES_DEFINED)
  #undef SUPERLU_ALIASES_DEFINED
  #undef ScalePermstruct_t
  #undef alias2..
...
#else 
  #define SUPERLU_ALIASES_DEFINED
  #define ScalePermstruct_t dScalePermstruct_t    /* zScalePermstruct_t  in  superlu_zdefs.h */
  #define alias2
...
#endif
 ``` 
A bit messy...

**Comment by keitat** [comment]
> This type of alias does not work if the user wants to have both real and complex *in one executable at the same time. *That is, they want both #include <superlu_ddefs.h> #include <superlu_zdefs.h> in the same code. The old way seems to work for some compilers. But at least some compilers break, complaining conflicts or multiple definitions, and cannot proceed (not just warnings). That's why I split into separate names. As I mentioned, in the future, when we want to support more mixed precision / mixed type codes, it will be even more messy.
> […](#)

Amesos (old generation library) is not template based.  This can live with old SuperLU_DIST solution. 
The new release is favorable for Amesos2 as we use type mapping and precision specific namespaces.   

@xiaoyeli 
Another question, we, Trilinos team, view the 6.3 release as a major version-up.  Do you still treat it as a minor version-up?   I am curious with the new features of the 7.0 release.  This also makes me consider our community rules on  version-ups.

**Comment by balay** [comment]
| Amesos (old generation library) is not template based. This can live with old SuperLU_DIST solution.

What is this solution that trilinos is advocating?

Note: trilinos enables both amesos and amesos2 in spack so both should be working with the the same superlu-dist version.

As of now - the solution I have is disable superlu-dist interface from trilinos.

**Comment by jedbrown** [comment]
It would save everyone time if Amesos were to apply the compatibility patch. It's simple, incurs no internal maintenance costs, and doesn't disrupt users. The [Amesos website](https://trilinos.github.io/amesos.html) doesn't say it's deprecated [1]; if you need to deprecate it because you don't want to support building Amesos and Amesos2 in the same application (due to conflicting dependencies), that's a decision to make outside this thread and not in the last week of an xSDK release.

[1] The version numbers on the Amesos website are way out of date.

**Comment by xiaoyeli** [comment]
> > This type of alias does not work if the user wants to have both real and complex *in one executable at the same time. *That is, they want both #include <superlu_ddefs.h> #include <superlu_zdefs.h> in the same code. The old way seems to work for some compilers. But at least some compilers break, complaining conflicts or multiple definitions, and cannot proceed (not just warnings). That's why I split into separate names. As I mentioned, in the future, when we want to support more mixed precision / mixed type codes, it will be even more messy.
> > […](#)
> 
> Amesos (old generation library) is not template based. This can live with old SuperLU_DIST solution.
> The new release is favorable for Amesos2 as we use type mapping and precision specific namespaces.
> 
> @xiaoyeli
> Another question, we, Trilinos team, view the 6.3 release as a major version-up. Do you still treat it as a minor version-up? I am curious with the new features of the 7.0 release. This also makes me consider our community rules on version-ups.

@keitat That was my mistake. It should have been bumped up to v7.xxx, because it involves interface change.
I was reserving v7 for the new code with 3D communication-avoiding algorithms. It's a lot more involved to put in production level of release.  We are still working on it, may not make it the the upcoming xSDK release.

**Comment by balos1** [comment]
When installing via spack sundials now includes a patch for our 5.4.0 release that allows it to work with superlu-dist 6.3.0+.

**Comment by balay** [comment]
@keitat @jwillenbring  I ran a build with trilinos+superlu-dist with the patch from https://github.com/trilinos/Trilinos/pull/8137 - and that went through.

**Comment by balos1** [comment]
I believe this is solved for all the packages.
